clifford/numba/_multivector.py

```diff
@@ -57,15 +56,16 @@ def _numba_type_(self):
 
     cache = layout_type._cache
     dt = self.value.dtype
+    C_contig = self.value.flags.C_CONTIGUOUS
 
     # now use the dtype to key that cache.
     try:
-        return cache[dt]
+        return cache[(dt, C_contig)]
     except KeyError:
         # Computing and hashing `dtype_type` is slow, so we do not use it as a
         # hash key. The raw numpy dtype is much faster to use as a key.
-        dtype_type = _numpy_support.from_dtype(dt)
-        ret = cache[dt] = MultiVectorType(layout_type, dtype_type)
+        dtype_type = _typeof_ndarray(self.value, None)
+        ret = cache[(dt, C_contig)] = MultiVectorType(layout_type, dtype_type)
```

Comment:
I'm starting to think that using `_typeof_ndarray` is a mistake after all - the only way to create a safe cache is to use it as the key, and that just costs way too much. For `a = np.ones(32)`, I measure:
```python
# the cost of a multivector addition
In [55]: %timeit a + a
1.16 µs ± 263 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
```python
# the current cache key
In [82]: d = {a.dtype: None}

In [83]: %timeit d[a.dtype]
208 ns ± 16.9 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [84]: %timeit d[a.dtype]
204 ns ± 36.5 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
```python
# your cache key
In [69]: d = {(a.dtype, a.flags.c_contiguous): None}

In [70]: %timeit d[(a.dtype, a.flags.c_contiguous)]
466 ns ± 129 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [71]: %timeit d[(a.dtype, a.flags.c_contiguous)]
491 ns ± 156 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
```python
# a cache key that includes the memory order
In [72]: d = {(a.dtype, map_layout(a)): None}

In [73]: %timeit d[(a.dtype, map_layout(a))]
716 ns ± 133 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [74]: %timeit d[(a.dtype, map_layout(a))]
824 ns ± 160 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
```python
# the full cache key
In [75]: d = {_typeof_ndarray(a, None): None}

In [76]: %timeit d[_typeof_ndarray(a, None)]
22.3 µs ± 926 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [77]: %timeit d[_typeof_ndarray(a, None)]
36.4 µs ± 7.76 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

I think our options are:

* Support _only_ C-order multivectors. Frankly, this is the most common case, and the cost of making a C-order copy is probably lower than the dispatch overhead. We don't need to change our caching if we do this. I don't know what the behavior is if you have a non-C order multivector. Ideally it silently copies, but it might refuse to compile.
* Support C-order and A-order multivectors (via your cache key). We should then construct the appropriate `Array` type ourselves instead of calling `_typeof_ndarray`, as that's the only way to ensure our cache is truthful. There's no point paying for F-order multivectors, as A is a superset of F that's just less efficient.

Reply:
Nice investigation, I think the C-only option makes sense (and is the most performant) as long as there are no specific really desirable use cases for A ordered arrays?